src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp

```diff
@@ -58,51 +54,32 @@ struct InitializeResidualMagnitude {
     using residual_magnitude_tag = db::add_tag_prefix<
         LinearSolver::Tags::Magnitude,
         db::add_tag_prefix<LinearSolver::Tags::Residual, fields_tag>>;
+    using initial_residual_magnitude_tag =
+        db::add_tag_prefix<LinearSolver::Tags::Initial, residual_magnitude_tag>;
 
-    db::mutate<residual_magnitude_tag>(
+    db::mutate<residual_magnitude_tag, initial_residual_magnitude_tag>(
         make_not_null(&box), [residual_magnitude](
                                  const gsl::not_null<double*>
-                                     stored_residual_magnitude) noexcept {
-          *stored_residual_magnitude = residual_magnitude;
+                                     local_residual_magnitude,
+                                 const gsl::not_null<double*>
+                                     initial_residual_magnitude) noexcept {
+          *local_residual_magnitude = *initial_residual_magnitude =
+              residual_magnitude;
         });
 
-    // When more sophisticated convergence criteria are implemented we may also
-    // want to check for convergence here.
+    // Determine whether the linear solver has already converged. This invokes
+    // the compute item.
+    const auto& has_converged = db::get<LinearSolver::Tags::HasConverged>(box);
+
+    if (UNLIKELY(has_converged and
+                 static_cast<int>(get<::Tags::Verbosity>(box)) >=
+                     static_cast<int>(::Verbosity::Quiet))) {
+      Parallel::printf(MakeString{} << has_converged);
```

Comment:
`Parallel::printf("%s", has_converged)`  There are several other similar bits of code in this PR, which should also be changed.

The important part of this change is the `"%s"`.  The fact that you can (I think) avoid the `MakeString` is a side benefit.  If the string form of `has_converged` happens to contain a "%" then the current code may segfault.  The first argument to `printf` should (almost) always be a string literal.  (This is GCC's `-Wformat-security` warning for the C `printf`.)